src/Wavefront.OpenTracing.CSharp.SDK/WavefrontTracer.cs

```diff
@@ -0,0 +1,219 @@
+            ///     provided configuration.
+            /// </summary>
+            /// <returns>A <see cref="WavefrontTracer"/>.</returns>
+            public WavefrontTracer Build()
+            {
+                if (string.IsNullOrEmpty(source))
+                {
+                    source = GetDefaultSource();
+                }
+                if (reporter == null)
+                {
+                    reporter = new ConsoleReporter(source);
+                }
+                return new WavefrontTracer(reporter, tags);
+            }
+
+            private static string GetDefaultSource()
+            {
+                try
+                {
+                    return Dns.GetHostEntry("LocalHost").HostName;
+                }
+                catch (Exception)
+                {
+                    return "wavefront-tracer";
+                }
+            }
+        }
+
+        private WavefrontTracer(IReporter reporter, IList<KeyValuePair<string, string>> tags)
+        {
+            ScopeManager = new AsyncLocalScopeManager();
+            this.reporter = reporter;
+            Tags = tags;
+            // TODO: figure out sampling
+        }
+
+        /// <inheritdoc />
+        public IScopeManager ScopeManager { get; }
+
+        /// <inheritdoc />
+        public ISpan ActiveSpan => ScopeManager.Active?.Span;
+
+        internal IList<KeyValuePair<string, string>> Tags { get; }
+
+        /// <inheritdoc />
+        public ISpanBuilder BuildSpan(string operationName)
+        {
+            return new WavefrontSpanBuilder(operationName, this);
+        }
+
+        /// <inheritdoc />
+        public void Inject<TCarrier>(
+            ISpanContext spanContext, IFormat<TCarrier> format, TCarrier carrier)
+        {
+            var propagator = registry.Get(format);
+            if (propagator == null)
+            {
+                throw new ArgumentException("invalid format: " + format);
+            }
+            propagator.Inject((WavefrontSpanContext)spanContext, carrier);
+        }
+
+        /// <inheritdoc />
+        public ISpanContext Extract<TCarrier>(IFormat<TCarrier> format, TCarrier carrier)
+        {
+            var propagator = registry.Get(format);
+            if (propagator == null)
+            {
+                throw new ArgumentException("invalid format: " + format);
+            }
+            return propagator.Extract(carrier);
+        }
+
+        internal void ReportSpan(WavefrontSpan span)
+        {
+            // Reporter will flush it to Wavefront/proxy.
+            try
+            {
+                reporter.Report(span);
+            }
+            catch (IOException e)
+            {
+                logger.Log(LogLevel.Warning, "Error reporting span", e);
+            }
+        }
+
+        internal DateTimeOffset CurrentTimestamp()
+        {
+            return DateTimeOffset.UtcNow;

```

Comment:
Discussed offline regarding this. @hanwavefront will investigate whether `StopWatch` class is a better fit for computing elapsed time to account for clock drift/local clock changes.